internal/node/hybrid/certificate_validator.go

```
@@ -0,0 +1,148 @@
+package hybrid
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+)
+
+type ValidationErrorType int
+
+const (
+	ErrorNoCert ValidationErrorType = iota
+	ErrorCertFile
+	ErrorReadFile
+	ErrorInvalidFormat
+	ErrorClockSkewDetected
+	ErrorExpired
+	ErrorParseCA
+	ErrorInvalidCA
+)
+
+// ValidationError represents a kubelet certificate validation error with remediation information
+type ValidationError struct {
+	message             string
+	remediation         string
+	cause               error
+	isDateValidation    bool
+	noCert              bool
+	validationErrorType ValidationErrorType
+}
+
+func (e *ValidationError) ErrorType() ValidationErrorType {
+	return e.validationErrorType
+}
+
+func (e *ValidationError) Error() string {
+	if e.cause != nil {
+		return fmt.Sprintf("%s: %v", e.message, e.cause)
+	}
+	return e.message
+}
+
+func (e *ValidationError) Unwrap() error {
+	return e.cause
+}
+
+func (e *ValidationError) Remediation() string {
+	return e.remediation
+}
+
+func IsDateValidationError(err error) bool {
+	var v *ValidationError
+	return errors.As(err, &v) && v.isDateValidation
+}
+
+func IsNoCertError(err error) bool {
+	var v *ValidationError
+	return errors.As(err, &v) && v.noCert
+}
+
+// ValidateCertificate checks if there is an existing certificate and validates it against the provided CA
+func ValidateCertificate(certPath string, ca []byte) error {
+	if _, err := os.Stat(certPath); os.IsNotExist(err) {
+		// Return an error for no cert, but one that can be identified
+		return &ValidationError{
+			message:             "no certificate found",
+			noCert:              true,
+			validationErrorType: ErrorNoCert,
+		}
```

why do we swallow the original error? that one probably has the path of the file we are trying to read

---

This is from existing code - [kubelet_cert_validator.go](https://github.com/vikash-s1/eks-hybrid/blob/main/internal/node/hybrid/kubelet_cert_validator.go).